metadata-ingestion/src/datahub/sql_parsing/tool_meta_extractor.py

```diff
@@ -0,0 +1,96 @@
+import json
+import logging
+from dataclasses import dataclass, field
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+from typing_extensions import Protocol
+
+from datahub.ingestion.api.report import Report
+from datahub.metadata.urns import CorpGroupUrn, CorpUserUrn
+from datahub.utilities.stats_collections import int_top_k_dict
+
+UrnStr = str
+
+logger = logging.getLogger(__name__)
+
+
+class QueryLog(Protocol):
+    """Represents Query Log Entry
+    expected by QueryMetaExractor interface
+    """
+
+    query_text: str
+    user: Optional[Union[CorpUserUrn, CorpGroupUrn]]
```

Comment:
is the idea that both preparsed queries and observed queries will match this interface?

Reply:
Yes, ultimately, that's the goal . Anything that has "query", "user" and "extra_info" can use this. 
ObservedQuery and PreparsedQuery share these concepts but not named/typed similarly. Refractoring can be done to achieve this, however currently not done as add_observed_query path ultimately invokes add_preparsed_query path. So it suffices for only PreparsedQuery to follow this protocol.